Fixing bugs in the application process prior to opening the form

viktorcsimma opened the pull request:
In the end, I started it based on `development`; some things have changed in it that I needed.

Largely went through @horcsinbalint's suggestions on 4th July:

```
Sziasztok!
Gyorsan próbáltam átfutni az Urán felvételivel kapcsolatos kódját. Nem voltam túl részletes, néhány probléma, amit találtam:
- app/Http/Controllers/Auth/RegisterController.php:53 rossz
- resource/views/user/update_tenant_status.blade.php:30 rossz
- felvételiztetőknek nem listázza a felvételizőket
- a határidő és a lezárás néhány dolognál csak a user interface-en látható, a POST requesteket kezelő kód nem minden esetben ellenőrzi a határidőt (profilképpel, személyes adatokkal, tanulmányi adatokkal, ALFONSÓ-val, nyelvvizsgával kapcsolatos dolgokat)
- akinek jogosultsága van egy adott felhasználót megtekinteni, annak jogosultsága van módosítani az adatokat is. Vagyis felvételiztetők módosíthatják felvételizők előbb említett adatait
- hosszú inputokat az adatbázis utasít el, egy 500-as kód társaságában, nincs validálva az input (pl.: app/Http/Controllers/Auth/ApplicationController.php:274)
- pályázónak profilképével a CSP miatt nincs lehetősége XSS-t végrehajtani, ha SVG-t tölt fel, ha jól tudom
- ha létezik már a felhasználó által meghatározott neptun kódú felhasználónak WiFije, akkor az app/Http/Controllers/Auth/ApplicationController.php:330 egy gyönyörű errort dob, és nem kap internethozzáférést az illető
- app/Http/Controllers/Auth/ApplicationController.php:317-318 sorrendje nem túl logikus olyan szempontból, hogy előbb a DB-ből a referenciát lenne talán célszerűbb kiszedni, mert így ha rosszkor száll el a rendszer, a hivatkozás megmaradhat egy nem létező fájlra
- app/Http/Controllers/Auth/ApplicationController.php:194-205 elég sus. Bárki, aki részt vesz a felvételi folyamatban módosíthatja bárkinél a megjegyzést, még ha nem is nézhetné meg az érintett jelentkezést (?!). És módosíthatja a státuszát is a jelentkezésnek (?!). Jelen kontextusban részt vesz a felvételi folyamatban (app/Policies/ApplicationFormPolicy.php:53-64): Role::SECRETARY, Role::DIRECTOR, Role::WORKSHOP_ADMINISTRATOR, Role::WORKSHOP_LEADER, Role::APPLICATION_COMMITTEE_MEMBER, Role::STUDENT_COUNCIL => Role::STUDENT_COUNCIL_LEADERS, Role::AGGREGATED_APPLICATION_COMMITTEE_MEMBER
- abnormálisan keskeny és magas profilképek egész szórakoztatóvá teszik a pályázó profilját
```

My comments:
- The first two have already been fixed in `development`.
- As I checked it out, it does not let the user upload an SVG anyway; neither as an avatar nor as an attachment. The textual fields seem to handle XSS correctly.
- On the uniqueness of internet accesses: my solution was not to let the user provide a Neptun code which already exists. _(At least they are not going to create multiple accounts...)_ I have not made the column itself unique so as not to mess up with already-existing duplicate accounts.
- Notes can now be modified by those who can view the profile, apart from the applicant themselves.

For finding further issues, I am going to ask Students' Council members to test the form on `staging`, if you agree. It looks good to me right now, but I am also going to fill it some more times, of course.

viktorcsimma committed:
Addressing the previous spottings of @horcsinbalint

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 8317488 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/572

horcsinbalint commented:
Have you checked on this? `nem menti el, ha valakinek a jelentkezésének a státusza módosítva van a "Felvételi" lapon`

viktorcsimma commented:
> Have you checked on this? `nem menti el, ha valakinek a jelentkezésének a státusza módosítva van a "Felvételi" lapon`

You are right, I've noticed that but I've passed... I have not found any solutions yet, but I will try again tomorrow. Until then, let this remain a draft.

horcsinbalint commented:
> > Have you checked on this? `nem menti el, ha valakinek a jelentkezésének a státusza módosítva van a "Felvételi" lapon`
> 
> You are right, I've noticed that but I've passed... I have not found any solutions yet, but I will try again tomorrow. Until then, let this remain a draft.

I checked and the bug was introduced in #490 where laravel and livewire was updated. I did not look into it but this problem probably has to do something with the update of livewire ([resources/views/auth/application/application.blade.php:20](https://github.com/EotvosCollegium/mars/blob/7d0649f876cbf4ba58cf0e8999a896fe75dbf8da/resources/views/auth/application/application.blade.php#L20)).

horcsinbalint commented:
Probably off-topic: there was the feature request of adding some kind of text for applicants about special additional requirements for different workshops. (Like uploading a CV in a specific language, the requirement to also solve a quiz, etc.) Do you plan on incorporating it into this patch or are you planning on making a different PR for it?

viktorcsimma committed:
Rewriting status updater

viktorcsimma committed:
Updating a test case for a new validation rule

viktorcsimma commented:
> > > Have you checked on this? `nem menti el, ha valakinek a jelentkezésének a státusza módosítva van a "Felvételi" lapon`
> > 
> > 
> > You are right, I've noticed that but I've passed... I have not found any solutions yet, but I will try again tomorrow. Until then, let this remain a draft.
> 
> I checked and the bug was introduced in #490 where laravel and livewire was updated. I did not look into it but this problem probably has to do something with the update of livewire ([resources/views/auth/application/application.blade.php:20](https://github.com/EotvosCollegium/mars/blob/7d0649f876cbf4ba58cf0e8999a896fe75dbf8da/resources/views/auth/application/application.blade.php#L20)).

Now it's done; I've rewritten it:) I'll push that soon.

> Probably off-topic: there was the feature request of adding some kind of text for applicants about special additional requirements for different workshops. (Like uploading a CV in a specific language, the requirement to also solve a quiz, etc.) Do you plan on incorporating it into this patch or are you planning on making a different PR for it?

I think I'll open a separate issue for that.

kdmnk commented:
Be aware that this will have conflicts with #560. We need to make sure to handle merge conflicts properly to not lose changes.

viktorcsimma committed:
Fixing a bug in the Neptun code checker

viktorcsimma committed:
Adding instruction changes by @kdmnk

viktorcsimma committed:
Merge development into application-hotfixes

viktorcsimma committed:
Solving nested localization rule without dots in keys

viktorcsimma committed:
Following suggestion of phpstan on an is_null call

viktorcsimma committed:
Caching periodic event start and end dates

viktorcsimma commented on app/Policies/UserPolicy.php:
```diff
@@ -114,6 +115,25 @@
         return false;
     }
 
+    /**
+     * Whether one can edit the user's educational or personal information.
+     *
+     * @param User $user
+     * @param User $target
+     * @return bool
+     */
+    public function updateInformation(User $user, User $target): bool
+    {
+        if (!$target->verified && $target->isCollegist()) {
+            // that means, they are applicants
+            $endDate = app(\App\Http\Controllers\Auth\ApplicationController::class)->getEndDate();
+            return (is_null($endDate) || Carbon::now() < $endDate)
+            && ($user->id == $target->id);
+        } else {
+            return $this->view($user, $target);
+        }
+    }
```
It can be expected not to throw. If still, let it fail.

viktorcsimma commented on app/Http/Controllers/Auth/ApplicationController.php:
```diff
@@ -189,10 +189,11 @@
      */
-    public function editApplication(Request $request): RedirectResponse
+    public function editApplicationNote(Request $request): RedirectResponse
     {
-        $this->authorize('viewSome', ApplicationForm::class);
         $application = ApplicationForm::findOrFail($request->input('application'));
+        $this->authorize('editNote', $application);
+
```
That will be made obsolete by #574. But nice suggestion otherwise.

viktorcsimma committed:
Improving validation in ApplicationController

viktorcsimma commented on app/Http/Controllers/Auth/ApplicationController.php:
```diff
@@ -253,35 +254,44 @@ public function finalizeApplicationProcess()
         ]);
+        $validatedData['accommodation'] = $request->input('accommodation') === "on";
+
+        ApplicationForm::updateOrCreate(['user_id' => $user->id], $validatedData);
```
Done;)

viktorcsimma commented on app/Policies/UserPolicy.php:
```diff
@@ -114,6 +115,25 @@ public function view(User $user, User $target): bool
+     * @param User $target
+     * @return bool
+     */
+    public function updateInformation(User $user, User $target): bool
+    {
+        if (!$target->verified && $target->isCollegist()) {
+            // that means, they are applicants
+            $endDate = app(\App\Http\Controllers\Auth\ApplicationController::class)->getEndDate();
+            return (is_null($endDate) || Carbon::now() < $endDate)
+            && ($user->id == $target->id);
+        } else {
+            return $this->view($user, $target);
+        }
+    }
```
Done;)

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 6691d64 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/572

kdmnk commented on app/Http/Controllers/Auth/ApplicationController.php:
```diff
@@ -253,35 +254,45 @@ public function finalizeApplicationProcess()
-            'accommodation' => $request->input('accommodation') === "on",
-            'present' => $request->input('present')
+        $validatedData = $request->validate([
+            'graduation_average' => 'required|numeric|min:1|max:5',
```
I already implemented this in one of the other PRs.

kdmnk commented on app/Http/Controllers/Auth/ApplicationController.php:
```diff
@@ -253,35 +254,45 @@ public function finalizeApplicationProcess()
         if ($request->input('status') == 'resident') {
             $user->setResident();
-        } elseif ($request->input('status') == 'extern') {
+        } else {
```
Overridden by one of the other PRs.

kdmnk commented on app/Http/Controllers/Secretariat/UserController.php:
```diff
@@ -171,6 +171,11 @@ public function updateEducationalInformation(Request $request, User $user): Redi
 
+        // whether Neptun code is unique
+        if (EducationalInformation::where('neptun', $request->neptun)->where('user_id', '<>', $user->id)->exists()) {
+            return redirect()->back()->with('message', 'A megadott Neptun-kód már létezik! Ha a kód az Öné, lépjen be a korábbi fiókjával.');
```
->with('error')...
It will display a warning sign and the toast have a longer display period.

kdmnk commented on app/Livewire/ApplicationStatusUpdate.php:
```diff
@@ -7,21 +7,27 @@
+    /**
+     * Updates the status of the application and flashes a confirmation message.
+     */
+    public function setStatus($status)
```
this is also overridden by https://github.com/EotvosCollegium/mars/pull/574

kdmnk commented on app/Policies/UserPolicy.php:
```diff
@@ -114,6 +115,25 @@ public function view(User $user, User $target): bool
+    public function updateInformation(User $user, User $target): bool
+    {
+        if (!$target->verified && $target->isCollegist()) {
+            // that means, they are applicants
```
not at all.
- a verified user can still be an applicant
- new applicants don't have collegist role from now on

Editing data for applicants is not yet implemented anyways, I would just delete this for now.

kdmnk commented on app/Utils/HasPeriodicEvent.php:
```diff
@@ -96,6 +97,10 @@ final public function updatePeriodicEvent(Semester $semester, Carbon $start_date
             $event->save();
             $event->refresh();
+
+            Cache::put($this->underlyingControllerName . '_start_date', $start_date);
```
I see why do you want caching here, but I would rather make it in a separate issue or in https://github.com/EotvosCollegium/mars/pull/555 as it will also rename underlyingControllerName.

kdmnk commented on app/Utils/HasPeriodicEvent.php:
```diff
@@ -155,15 +160,23 @@ final public function isExtended(): bool
     final public function getStartDate(): ?Carbon
     {
-        return $this->periodicEvent()?->startDate();
+        return Cache::remember(
```
same as before

kdmnk commented on app/Utils/HasPeriodicEvent.php:
```diff
@@ -155,15 +160,23 @@ final public function isExtended(): bool
     final public function getEndDate(): ?Carbon
     {
-        return $this->periodicEvent()?->endDate();
+        return Cache::remember(
```
same as before

kdmnk commented on resources/lang/hu/validation.php:
```diff
@@ -64,6 +66,10 @@
         'alfonso_language' => 'Alfonsó nyelv',
         'alfonso_desired_level' => 'Alfonsó szint',
+        'semester_average' => array_fill(0, 10, 'korábbi szemeszter átlaga'), // from semester_average.0 to semester_average.9
+        'question_1' => "\"Honnan hallott a Collegiumról?\" kérdés",
```
this is also implemented in one of my other PRs 🙃 . But maybe not all. Compare when merging.

kdmnk commented on resources/views/auth/application/questions.blade.php:
```diff
@@ -103,20 +103,27 @@
                     <x-input.textarea id="question_3"
-                                      text="Tervez-e tovább tanulni a diplomája megszerzése után? Milyen tervei vannak az egyetem után?"
+                                      text="Tervez-e továbbtanulni a diplomája megszerzése után? Milyen tervei vannak az egyetem után?"
+                                      maxlength="1000"
```
have you tested this in the ui?

viktorcsimma replied:
Yes, I did:) the only problem was that it did not count newlines the same way as the validation rule did.

kdmnk commented on resources/views/auth/application/status_update_component.blade.php:
```diff
@@ -1,19 +1,33 @@
 <div class="right">
-    @foreach (\App\Models\ApplicationForm::STATUSES as $st)
-    @if($st != \App\Models\ApplicationForm::STATUS_IN_PROGRESS || user()->isAdmin())
+    <div>
```
also overridden by https://github.com/EotvosCollegium/mars/pull/574.
What was the point here anyways?

viktorcsimma replied:
For some reason, it did not update the status for me and I rewrote the whole thing :sweat_smile:

kdmnk reviewed (commented):
The code itself is ok, but we have a tons of merge conflicts, also with moved files so the annoying kind. How should we proceed?
We can try to merge the others first and I can try to solve the conflicts on this one but to prepare that it would be nice to revert some changes here.

viktorcsimma commented:
I think we should merge the others first. I'll be grateful if you could try to solve conflicts:) but I can try it too.

kdmnk commented:
I merged the others. I will try to solve this tomorrow sometime, but I think the easiest is to just redo the changes in a new branch. Do you want to give it a go?

viktorcsimma commented:
> I merged the others. I will try to solve this tomorrow sometime, but I think the easiest is to just redo the changes in a new branch. Do you want to give it a go?

That's exactly what has crossed my mind a few minutes ago:D I'll try that; let's close this until then.
